### PR TITLE
Fix `LazyEvaluatedKernelTensor._unsqueeze_batch`

### PR DESCRIPTION
**Pull request by valtron**
This is in response to #1813, which caused some bugs as the indexing code in lazy tensors did not support `None` indexing.

Addresses:
- #1827
- pytorch/botorch#980
- pytorch/botorch#976

Tests are passing locally* for GPyTorch (with this patch) and BoTorch (with patch pytorch/botorch#976).

*Except for some CUDA tests that failed before, and are unrelated

### CODE REVIEW

File: gpytorch/lazy/cat_lazy_tensor.py [hunk 1]
```diff
@@ -185,7 +185,18 @@ def _get_indices(self, row_index, col_index, *batch_indices):
 
     def _getitem(self, row_index, col_index, *batch_indices):
         indices = [*batch_indices, row_index, col_index]
-        cat_dim_indices = indices[self.cat_dim]
+        cat_dim = self.cat_dim % len(self.shape)
```
Comment (Balandat) [inline gpytorch/lazy/cat_lazy_tensor.py:188]: Is this something we'd want to do at a higher level (e.g. upon init)?

File: gpytorch/lazy/cat_lazy_tensor.py [hunk 2]
```diff
@@ -185,7 +185,18 @@ def _get_indices(self, row_index, col_index, *batch_indices):
...
+            if idx is None:
+                cat_dim += 1
+                continue
+            if i == cat_dim:
+                break
+        cat_dim_indices = indices[cat_dim]
+        out_cat_dim = cat_dim - sum(isinstance(idx, int) for idx in indices[:cat_dim])
+        num_tensor_indices = sum(torch.is_tensor(idx) for idx in indices)
+        if num_tensor_indices > 0:
+            out_cat_dim -= num_tensor_indices - 1
```
Comment (Balandat) [inline gpytorch/lazy/cat_lazy_tensor.py:199]: I *think* I understand what you're doing here, but it's not at all easy to follow. Could you please add some comments here to explain what's going on?

**Review by Balandat — COMMENTED** [review]
Awesome, thanks for submitting this fix. I have some questions as to exactly what you're doing in the code (see inline). I also think that @gpleiss of @jacobrgardner better take a look at this as well since this again goes into the indexing weeds.

File: gpytorch/lazy/cat_lazy_tensor.py (hunk 1 above)
Comment (valtron) [reply gpytorch/lazy/cat_lazy_tensor.py:188]: `cat_dim` is actually normalized to a negative index in the init, and here I'm making it positive again for convenience.

When I originally looked at this code, I thought the reason `cat_dim` is stored as a negative index is to avoid having to do adjustments to it when when indexing through the cat... but on second thought, adjustments would still be needed, e.g.:

```
M = zeros(size = (2, 3, 4, 5, 6))
C = cat([M, M], dim = -4)
C[:,:,0,:,:] == cat([M[:,:,0,:,:], M[:,:,0,:,:]], dim = -3) # -4 -> -3
```

(I haven't had time to take a deeper look since making this PR.)

File: gpytorch/lazy/cat_lazy_tensor.py (hunk 2 above)
Comment (valtron) [reply gpytorch/lazy/cat_lazy_tensor.py:199]: The first part (189 - 194) is adjusting `cat_dim` to take into account unsqueezed dimensions. Line 196 is adjusting the `cat_dim` of the returned lazy cat tensor to take into account dimensions that were removed because of `int` indexing. And lines 197 - 199 adjust it for tensor indexes [1]: if `indices` contains `n > 0` tensor indices, those dimensions are removed and a dimension is added to the beginning [2].

... except turns out indexing is more complicated than I originally thought.
[1] Firstly, here I'm not handling the case when the shape of the tensor index has len() > 1. The number of dimensions added is actually the number of dimensions of (the broadcast of) all the tensor indices. 😱
[2] Apparently (this is really weird), if _all_ tensor indices are in adjacent dimensions (`int` indexing is a type of tensor indexing, just with shape `()`), the new dimensions are added to in place of the removed dimensions. If they're not adjacent, the new dimensions are added to the beginning: 😱😱
```python
import numpy as np
A = np.zeros((4,5,6,7,8))
idxs = [[0,0],[0,0],[0,0]]
print(A.shape)
# (4, 5, 6, 7, 8)
print(A[:,idxs,idxs,idxs,:].shape)
# (4, 3, 2, 8)
print(A[:,idxs,:,idxs,idxs].shape)
# (3, 2, 4, 6)
```

So, this will take more work to work properly. I'll leave adding comments until the end.

File: gpytorch/lazy/lazy_tensor.py [hunk 3]
```diff
@@ -2092,6 +2092,9 @@ def __getitem__(self, index):
         index = tuple(torch.tensor(idx) if isinstance(idx, list) else idx for idx in index)
         index = tuple(idx.item() if torch.is_tensor(idx) and not len(idx.shape) else idx for idx in index)
 
+        # Increase ndimension by the number of dimensions unsqueezed via `None` indexing
+        ndimension += sum(idx is None for idx in index)
+
```
Comment (gpleiss) [inline gpytorch/lazy/lazy_tensor.py:2097]: I think here it would make sense to call the `self._unsqueeze_batch` for each of the `None` indexes. I think this should be all there is to this PR?

Comment (valtron) [reply gpytorch/lazy/lazy_tensor.py:2097]: Because of https://github.com/cornellius-gp/gpytorch/pull/1813/files that would be infinite recursion. But you are right, in that rather than trying to support `None` indexing in lazy tensors, it might be easier (with smaller chance of breaking things) to implement a non-evaluating `LazyEvaluatedKernelTensor._unsqueeze_batch` directly. (The reason I punted it off to `__getitem__` was because I didn't know `None` indexing wasn't supported.) I tried something basic [1] just on the off chance it worked, it doesn't though. Kind of stuck right now...

[1] 
```python
indexes = (slice(None),) * dim + (None,)
return self.__class__(self.x2, self.x1, kernel=self.kernel[indexes], ...)
```

Comment (Balandat) [reply gpytorch/lazy/lazy_tensor.py:2097]: @gpleiss, @wjmaddox any further thoughts here? What's currently on master still breaks the botorch HighOrderGP tests, which I'd like to fix...

**Comment by valtron** [comment]
@Balandat @gpleiss Hey, I forgot about this for a few days. This new way of doing `_unsqueeze_batch` seems to work.

**Review by Balandat — COMMENTED** [review]
This lgtm, also it fixes the issues on the botorch end. If you can just fix the formatting issue with black I think this is good to go.

**Comment by Balandat** [comment]
The difficulties with black here are probably b/c the pre-commit hooks are using a pretty old version and what you have locally is different. Let me see if I can update this in another PR.

**Comment by valtron** [comment]
I tried using 19.10b0 as well, no luck -_-

**Comment by Balandat** [comment]
I just put up #1842 that changes a whole swath of things and upgraded versions, with that the 21.11b1 formatting should work.

**Review by gpleiss — APPROVED** [review]

**Comment by gpleiss** [comment]
LGTM! Let me see if I can fix the listing issue

**Comment by valtron** [comment]
Thanks @gpleiss! I'm curious, what did you do? For me `black` still reformats it to what github's CI rejects.

```
~/Desktop/a/gpytorch ((HEAD detached at origin/patch-1))
$ py -m black --version
python -m black, version 19.10b0

~/Desktop/a/gpytorch ((HEAD detached at origin/patch-1))
$ py -m black -l 120 --target-version=py36 gpytorch/lazy/lazy_evaluated_kernel_tensor.py
reformatted gpytorch\lazy\lazy_evaluated_kernel_tensor.py
All done! ✨ 🍰 ✨
1 file reformatted.

~/Desktop/a/gpytorch ((HEAD detached at origin/patch-1))
$ git diff
diff --git a/gpytorch/lazy/lazy_evaluated_kernel_tensor.py b/gpytorch/lazy/lazy_evaluated_kernel_tensor.py
index 016edda9..5c0e3f4c 100644
--- a/gpytorch/lazy/lazy_evaluated_kernel_tensor.py
+++ b/gpytorch/lazy/lazy_evaluated_kernel_tensor.py
@@ -135,7 +135,13 @@ class LazyEvaluatedKernelTensor(LazyTensor):
             new_kernel = self.kernel.__getitem__(batch_indices)

         # Now construct a kernel with those indices
-        return self.__class__(x1, x2, kernel=new_kernel, last_dim_is_batch=self.last_dim_is_batch, **self.params,)
+        return self.__class__(
+            x1,
+            x2,
+            kernel=new_kernel,
+            last_dim_is_batch=self.last_dim_is_batch,
+            **self.params,
+        )
(etc.)
```

**Comment by Balandat** [comment]
I think this was merged failing, I fixed it as part of #1842
